File: java_membership_api/src/main/java/com/example/membership_api/controller/MembershipTierController.java
package com.example.membership_api.controller;

import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.annotations.AdminOnly;
import com.example.membership_api.model.MembershipTier;
import com.example.membership_api.service.MembershipTierService;
import com.example.membership_api.utils.ResponseUtil;

@RestController
@RequestMapping("/tiers")
public class MembershipTierController {

    @Autowired
    private MembershipTierService membershipTierService;

    @GetMapping("")
    public ResponseEntity<?> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage,
            @RequestParam(required = false) Boolean active,
            @RequestParam(required = false) Integer tierLevel) {

        // Build filter map
        Map<String, Object> filters = new java.util.HashMap<>();
        if (active != null) filters.put("isActive", active);
            
        Map<String, Object> data = membershipTierService.getAll(page, perPage, filters);
        return ResponseUtil.success(null, data);
    }

    @GetMapping("/{tierId}")
    public ResponseEntity<?> getOne(@PathVariable String tierId) {
        MembershipTier tier = membershipTierService.getById(tierId);
        if (tier == null) {
            return ResponseUtil.error("Tier not found", 404);
        }
        return ResponseUtil.success( null,tier);
    }

    @PostMapping("")
    @AdminOnly
    public ResponseEntity<?> create(@RequestBody Map<String, Object> membershipTier) {
        MembershipTier created = membershipTierService.create(membershipTier);
        return ResponseUtil.success("Tier created", created);
    }

    @PutMapping("/{tierId}")
    @AdminOnly
    public ResponseEntity<?> update(@PathVariable String tierId, @RequestBody Map<String, Object> membershipTier) {
        MembershipTier updated = membershipTierService.update(tierId, membershipTier);
        if (updated == null) {
            return ResponseUtil.error("Tier not found", 404);
        }
        return ResponseUtil.success("Tier updated", updated);
    }

    @DeleteMapping("/{tierId}")
    @AdminOnly
    public ResponseEntity<?> delete(@PathVariable String tierId) {
        boolean deleted = membershipTierService.softDelete(tierId);
        if (!deleted) {
            return ResponseUtil.error("Tier not found", 404);
        }
        return ResponseUtil.success("Deleted successfully");
    }
}

File: flask_api/app/controllers/membership_tier_controller.py
from flask import request
from app.services.membership_tier_service import MembershipTierService
from app.utils.response import success_response, error_response
from app.utils.decorators import admin_required


class MembershipTierController:

    @staticmethod
    def get_all():
        page = request.args.get("page", default=1, type=int)
        per_page = request.args.get("per_page", default=10, type=int)

        filters = {
            "is_active": request.args.get(
                "active", type=lambda v: v.lower() == "true" if v else None
            ),
            "tier_level": request.args.get("tier_level"),
        }

        filters = {k: v for k, v in filters.items() if v is not None}

        tiers, total = MembershipTierService.get_all(page, per_page, filters=filters)
        return success_response(
            "Membership Tiers retrieved",
            {
                "items": [t.to_dict() for t in tiers],
                "total": total,
                "page": page,
                "per_page": per_page,
            },
        )

    @staticmethod
    def get_one(tier_id):
        tier = MembershipTierService.get_by_id(tier_id)
        if not tier:
            return error_response("Tier not found", 404)
        return success_response("Tier fetched", tier.to_dict())

    @staticmethod
    @admin_required()
    def create():
        data = request.get_json()
        try:
            tier = MembershipTierService.create(data)
            return success_response("Tier created", tier.to_dict(), 201)
        except Exception as e:
            return error_response(str(e), 400)

    @staticmethod
    @admin_required()
    def update(tier_id):
        data = request.get_json()
        updated = MembershipTierService.update(tier_id, data)
        if not updated:
            return error_response("Tier not found", 404)
        return success_response("Tier updated", updated.to_dict())

    @staticmethod
    @admin_required()
    def delete(tier_id):
        success = MembershipTierService.soft_delete(tier_id)
        if not success:
            return error_response("Tier not found", 404)
        return success_response("Deleted successfully")
